### ISSUE TITLE
RangeError (index): Invalid value: Valid value range is empty: 0

### ISSUE THREAD

**Issue by zdongcoding** [shamblett/mqtt_client#229]
[ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: RangeError (index): Invalid value: Valid value range is empty: 0
E/flutter (12926): #0      List.[] (dart:core-patch/growable_array.dart:177:60)
E/flutter (12926): #1      SubscriptionsManager.confirmSubscription
package:mqtt_client/src/mqtt_client_subscriptions_manager.dart:176
E/flutter (12926): #2      MqttConnectionHandlerBase.messageAvailable
package:mqtt_client/…/connectionhandling/mqtt_client_mqtt_connection_handler_base.dart:220
E/flutter (12926): #3      _rootRunUnary (dart:async/zone.dart:1198:47)
E/flutter (12926): #4      _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter (12926): #5      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
E/flutter (12926): #6      CastStreamSubscription._onData (dart:_internal/async_cast.dart:85:11)
E/flutter (12926): #7      _rootRunUnary (dart:async/zone.dart:1198:47)
E/flutter (12926): #8      _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter (12926): #9      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
E/flutter (12926): #10     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
E/flutter (12926): #11     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
E/flutter (12926): #12     _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:127:11)
E/flutter (12926): #13     _WhereStream._handleData (dart:async/stream_pipe.dart:199:12)
E/flutter (12926): #14     _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:157:13)
E/flutter (12926): #15     _rootRunUnary (dart:async/zone.dart:1198:47)
E/flutter (12926): #16     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter (12926): #17     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
E/flutter (12926): #18     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
E/flutter (12926): #19     _DelayedData.perform (dart:async/stream_impl.dart:611:14)
E/flutter (12926): #20     _StreamImplEvents.handleNext (dart:async/stream_impl.dart:730:11)
E/flutter (12926): #21     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:687:7)
E/flutter (12926): #22     _rootRun (dart:async/zone.dart:1182:47)
E/flutter (12926): #23     _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter (12926): #24     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter (12926): #25     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter (12926): #26     _rootRun (dart:async/zone.dart:1190:13)
E/flutter (12926): #27     _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter (12926): #28     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter (12926): #29     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter (12926): #30     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter (12926): #31     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)

subAck.payload.qosGrants.length = 0

`   

     if (subAck.payload.qosGrants[0] == MqttQos.failure) {
       subscriptions.remove(topic);
       if (onSubscribeFail != null) {
         onSubscribeFail(topic);
        return false;
      }
   }

`

**Comment by shamblett** [comment]
The 3.1,1 spec for a suback says :
```
The payload contains a list of return codes. Each return code corresponds to a Topic Filter in the SUBSCRIBE Packet being acknowledged.
```
So the broker shouldn't send a suback with an empty QosGrants field in the payload, if the subscription has failed it should send code 0x80 to mark this. However this is an easy thing to catch, if a broker does this we have to assume the subscription has failed and act appropriately, I'll update the code.

**Comment by shamblett** [comment]
Fix in version 8.1.0
